3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,253 @@
+        // one, create an AmpAnalytics3pCreativeMessageRouter.
+        envelope[AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVES].forEach(creative => {
+          this.creativeMessageRouters_[creative.senderId] =
+            this.creativeMessageRouters_[creative.senderId] ||
+            new AmpAnalytics3pCreativeMessageRouter(this, this.sentinel_,
+              creative.senderId);
+          this.creativeMessageRouters_[creative.senderId]
+            .receiveNewCreativeMessage(creative);
+        });
+      });
+    this.iframeMessagingClient_.registerCallback(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENTS,
+      envelope => {
+        dev().assert(envelope[AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENTS],
+          'Received empty events envelope');
+        // Iterate through the array of events, dispatching each to the
+        // listener for the appropriate creative
+        envelope[AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENTS].forEach(event => {
+          dev().assert(this.creativeMessageRouters_[event.senderId],
+            'Received event message prior to new creative message.' +
+            ' Discarding.');
+          this.creativeMessageRouters_[event.senderId]
+            .receiveEventMessage(event);
+        });
+        this.flushQueues_();
+      });
+    this.sendReadyMessageToCreative_();
+  }
+
+  /**
+   * Gets a handle to the window (or testing object that mocks window)
+   * @returns {!Window}
+   */
+  getWindow() {
+    return this.win_;
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @private
+   */
+  sendReadyMessageToCreative_() {
+    this.iframeMessagingClient_.sendMessage(AMP_ANALYTICS_3P_MESSAGE_TYPE.READY,
+      {});
+  }
+
+  /**
+   * Sends a message back to the host window
+   * @param {string} type The type of message to send.
+   * @param {(ampAnalytics3pReadyMessage|ampAnalytics3pResponse)} opt_payload
+   * The payload of message to send.
+   */
+  sendMessage(type, opt_payload) {
+    this.iframeMessagingClient_.sendMessage(type, opt_payload);
+  }
+
+  /**
+   * Instructs each AmpAnalytics3pCreativeMessageRouter instances to send its
+   * queued messages to its listener (if if has a listener yet).
+   * @private
+   */
+  flushQueues_() {
+    Object.entries(this.creativeMessageRouters_).forEach(
+      entry => {
+        try {
+          const creativeMessageRouter = entry[1];
+          creativeMessageRouter.sendQueuedMessagesToListener();
+        } catch (e) {
+          dev().error(TAG_, 'Caught exception executing listener for ' +
+            entry[0] + ': ' + e.message);
+        }
+      });
+  }
+}
+
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!AmpAnalytics3pMessageRouter} parent
+   * @param {!string} creativeId
+   */
+  constructor(parent, creativeId) {
```

Comment:
Given you only construct this when you get the new creative message, seems to me you can just pass the extraData allowing it to be const.